witness/golang/internal/witness/db.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package witness
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// Database reads from and writes checkpoints to persistent storage.
+type Database struct {
+	db *sql.DB
+}
+
+// NewDatabase creates a new database, initializing it if needed.
+func NewDatabase(db *sql.DB) (*Database, error) {
+	_, err := db.Exec("CREATE TABLE IF NOT EXISTS chkpts (key BLOB, size INT, raw BLOB, PRIMARY KEY (key, size))")
+	if err != nil {
+		return nil, err
+	}
+	return &Database{db: db}, nil
+}
+
+// GetLatest reads the latest checkpoint written to the DB for a given log.
+func (d *Database) GetLatest(logID string) (*Chkpt, error) {
+	return d.getLatestChkpt(d.db.QueryRow, logID)
+}
+
+func (d *Database) getLatestChkpt(queryRow func(query string, args ...interface{}) *sql.Row, logID string) (*Chkpt, error) {
+	var maxChkpt Chkpt
+	row := queryRow("SELECT raw, size FROM chkpts WHERE key = ? ORDER BY size DESC LIMIT 1", logID)
+	if err := row.Err(); err != nil {
+		return nil, err
+	}
+	if err := row.Scan(&maxChkpt.Raw, &maxChkpt.Size); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, status.Errorf(codes.NotFound, "log %q not found", logID)
+		}
+		return nil, err
+	}
+	return &maxChkpt, nil
+}
+
+// SetCheckpoint writes the checkpoint to the DB for a given log, assuming
+// that the latest size is still what the caller thought it was.
+func (d *Database) SetCheckpoint(ctx context.Context, logID string, latestSize uint64, c *Chkpt) error {
+	tx, err := d.db.BeginTx(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("BeginTx: %w", err)
+	}
+	realLatest, err := d.getLatestChkpt(tx.QueryRow, logID)
+	if latestSize > 0 {
+		// If latest is non-zero check it's the same as the one in the DB.
+		if err != nil {
+			return fmt.Errorf("GetLatest: %w", err)
+		}
+		if latestSize != realLatest.Size {
+			return fmt.Errorf("latest checkpoint changed in the meantime")
+		}
+	} else {
+		// Otherwise check that there really isn't anything there.
+		if err == nil {
+			return fmt.Errorf("got latest=nil but there was a stored checkpoint")
```

Comment:
What if the stored checkpoint had size=0?

Reply:
Then things would be bad :).  I've added a comment about this assumption here but let me know if you think it belongs anywhere else too.

Reply:
I guess I don't see a compelling reason to change it - either way the first "real" checkpoint seen for a log is considered trust-on-first-use, so it's just a question of whether the log has entries or not at the point at which the witness is configured.  Since a feeder has no reason to follow a log until it's actively issuing new checkpoints it seems safe to assume that a log always has entries at the time the feeder adds it to the witness' list of known logs?